src/mbgl/style/conversion/filter.cpp

```diff
@@ -55,7 +55,8 @@ bool isExpression(const Convertible& filter) {
         return false;
         
     } else if (*op == "==" || *op == "!=" || *op == ">" || *op == ">=" || *op == "<" || *op == "<=") {
-        return arrayLength(filter) == 3 && (isArray(arrayMember(filter, 1)) || isArray(arrayMember(filter, 2)));
+        auto length = arrayLength(filter);
+        return (length == 3 || length == 4) && (isArray(arrayMember(filter, 1)) || isArray(arrayMember(filter, 2)));
```

Comment:
Apparently this code was ported [from GL JS code](https://github.com/mapbox/mapbox-gl-js/blob/master/src/style-spec/feature_filter/index.js) that will also need to be updated.

Reply:
Filed as https://github.com/mapbox/mapbox-gl-js/issues/6920

Reply:
Maybe slightly better would be:
```cpp
length != 3 || (isArray(arrayMember(filter, 1)) || isArray(arrayMember(filter, 2)))
```

Reply:
The reasoning being that the default should be "this is an expression", and we should scope the fallback to legacy filters as tightly as possible?

Reply:
Makes sense. Done.

Reply:
> The reasoning being that the default should be "this is an expression", and we should scope the fallback to legacy filters as tightly as possible?

Yeah, pretty much: sending as many inputs as possible through the expression parsing pipeline, since that will be more and more likely what users will be trying to write, and so expression-related validation messages are more likely to be useful